### ISSUE TITLE
java.lang.Error: MMF: Create device enumerator failed CoInitialize has not been called

### ISSUE THREAD

**Issue by nanguantong** [devopvoid/webrtc-java#43]
And then when use maven project, it's ok to run this jar, but error happened while running 
`List<AudioDevice> captureDevices = MediaDevices.getAudioCaptureDevices();`

the native code not called when WindowsAudioDeviceManager inited as following 
```
ComInitializer::ComInitializer() : initialized(false)
	{
		HRESULT hr = CoInitializeEx(NULL, COINIT_MULTITHREADED);

		if (hr != RPC_E_CHANGED_MODE) {
			THROW_IF_FAILED(hr, "Initialize COM failed");

			initialized = true;
		}
	}
```

in `WebRTCContext::initialize`
void WebRTCContext::initDeviceManagers()
	{
#ifdef _WIN32
		audioDevManager = std::make_unique<avdev::WindowsAudioDeviceManager>();
		videoDevManager = std::make_unique<avdev::WindowsVideoDeviceManager>();
#endif
      ......
}

in `WindowsAudioDeviceManager`
```
WindowsAudioDeviceManager::WindowsAudioDeviceManager() :
			deviceEnumerator()
		{
                       //////////////////////////////////////// not call ComInitializer /////////////////////////////////////////////
			HRESULT hr = CoCreateInstance(__uuidof(MMDeviceEnumerator), NULL, CLSCTX_ALL, __uuidof(IMMDeviceEnumerator), (void**)&deviceEnumerator);
			THROW_IF_FAILED(hr, "MMF: Create device enumerator failed");

			enumerateDevices(eAll);

			deviceEnumerator->RegisterEndpointNotificationCallback(this);
		}

```

**Comment by nanguantong** [comment]
@devopvoid can you please help to fix this bug and then release the maven or gradle jar packet to public repo?

**Comment by devopvoid** [comment]
For gradle please see How to fetch the jni libs with gradle #23.

**Comment by nanguantong** [comment]
> 

Yeah, got it.

**Comment by nanguantong** [comment]
> For gradle please see How to fetch the jni libs with gradle #23.

Yeah, got it.

**Comment by nanguantong** [comment]
> @devopvoid can you please help to fix this bug and then release the maven or gradle jar packet to public repo?

But this is realy a bug to call `CoCreateInstance `before COM `CoInitialize`, so hope to fix that.

**Comment by devopvoid** [comment]
This is not a bug. I'm not sure where you found the `ComInitializer`. 

1. WindowsAudio/VideoDeviceManager are based on Media Foundation.
2. WindowsAudioDeviceManager.h initializes MFInitializer.
3. MFInitializer calls MFStartup
4. Then WindowsAudioDeviceManager constructor is executed
5. Where CoCreateInstance works fine

**Comment by nanguantong** [comment]
> This is not a bug. I'm not sure where you found the `ComInitializer`.
> 
> 1. WindowsAudio/VideoDeviceManager are based on Media Foundation.
> 2. WindowsAudioDeviceManager.h initializes MFInitializer.
> 3. MFInitializer calls MFStartup
> 4. Then WindowsAudioDeviceManager constructor is executed
> 5. Where CoCreateInstance works fine


In `webrtc-jni\src\main\cpp\src\platform\windows\ComInitializer.cpp` call `HRESULT hr = CoInitializeEx(NULL, COINIT_MULTITHREADED);`

I just only call in my maven test project as following:

1. pom.xml:
        <dependency>
            <groupId>dev.onvoid.webrtc</groupId>
            <artifactId>webrtc-java</artifactId>
            <version>0.4.0-SNAPSHOT</version>
        </dependency> 
2. ![image](https://user-images.githubusercontent.com/12180902/142791735-ff73dc1c-dfef-442e-85e7-44bb688aab6c.png)

**Comment by devopvoid** [comment]
Your error actually points to `AudioProcessing`, not `MediaDevices.getAudioCaptureDevices()`. Another strange thing is, the code is executed in the `TestWebrtc.dispose` method where you`TestWebrtc.init` it.

Could you provide your test-class?

In the meantime you can try 0.5.0-SNAPSHOT.

**Comment by nanguantong** [comment]
![image](https://user-images.githubusercontent.com/12180902/142839786-a6ca43c5-e0ea-497a-9b4f-c3a9b682ebe6.png)

**Comment by nanguantong** [comment]
> Your error actually points to `AudioProcessing`, not `MediaDevices.getAudioCaptureDevices()`.
> 
> Could you provide your test-class?
> 
> In the meantime you can try 0.5.0-SNAPSHOT.

0.5.0-SNAPSHOT used ok.

**Comment by nanguantong** [comment]
> 

But 0.4.0 (SNAPSHOT) still has this problem.

**Comment by devopvoid** [comment]
I won't fix it for 0.4.0-SNAPSHOT since 0.4.0 is released. The next snapshot version is 0.5.0-SNAPSHOT.

**Comment by nanguantong** [comment]
> I won't fix it for 0.4.0-SNAPSHOT since 0.4.0 is released. The next snapshot version is 0.5.0-SNAPSHOT.

Yeh, no problem.
